# Documentation
There are two types of documentation to help you learn how to use CPTAC data for cancer research: tutorials and use cases. In general, tutorials are basic coding introduction to the API as well as conventions for storing and accessing the data. The use cases are short examples focused on a biological question and show practical uses of the API for biological discovery. Each use case works with a different combination of data types and explores meaningful cancer research hypotheses. 

Both the tutorials and use cases are written in Python using the interactive Jupyter notebooks. If you are unfamiliar with Jupyter, follow the instructions given <a href="https://jupyter.org/install">here</a> on the Jupyter website. You will then be able to run our tutorials as interactive, exploratory data analyses. If you want a no installation version, please visit our Binder site which hosts these notebooks [here](https://mybinder.org/v2/gh/PayneLab/cptac/master?filepath=%2Fdocs).


## Tutorials
- Tutorial 1: CPTAC data introduction
- Tutorial 2: Using pandas to work with cptac dataframes
- Tutorial 3: Joining dataframes with cptac
- Tutorial 4: Understanding multi-indexes
- Tutorial 5: How to keep up to date with new package and data releases
- Tutorial 6: Easy integration with R

## Use Cases
- Use Case 1: Comparing transcriptomics and proteomics
- Use Case 2: Correlation between clinical attributes
- Use Case 3: Associating clinical variables with omics data
- Use Case 4: How Do Mutations Affect Protein Abundance?
- Use Case 5: Gene Set Enrichment Analysis
- Use Case 6: Comparing Derived Molecular Data with Proteomics
- Use Case 7: Trans Genetics Effects
- Use Case 8: Outliers
- Use Case 9: Clinical Outcomes
- Use Case 10: Pathway diagram overlay
