{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial 1: CPTAC Data Introduction\n",
    "\n",
    "The National Cancer Instituteâ€™s Clinical Proteomic Tumor Analysis Consortium (CPTAC) is a national effort to accelerate the understanding of the molecular basis of cancer through the application of large-scale proteome and genome analysis, or proteogenomics. CPTAC generates comprehensive proteomics and genomics data from clinical cohorts, typically with ~100 samples per tumor type. The graphic below summarizes the structure of each CPTAC dataset. For more information, visit the [NIH website](https://proteomics.cancer.gov/programs/cptac). \n",
    "\n",
    "<img src=\"img/Graphical_Abstract.png\" alt=\"CPTAC cohort\" width=\"700\"/>\n",
    "\n",
    "This Python package makes accessing CPTAC data easy with Python code and Jupyter notebooks. The package contains several tutorials which demonstrate data access and usage. This first tutorial serves as an introduction to the data to help users become familiar with what is included and how it is presented."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Overview\n",
    "\n",
    "Our package provides data access in a Python programming environment. If you have not installed Python or have not installed the package, see our installation documentation [here](https://paynelab.github.io/cptac/#installation).\n",
    "\n",
    "Once we have the package installed and we're in our Python environment, we begin by importing the package with a standard Python import statement:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cptac"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "cptac data is broken down into datasets by cancer type. To view the available cancers and their abbreviations, call the `cptac.get_cancer_info()` function. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'brca': 'Breast invasive carcinoma',\n",
       " 'ccrcc': 'Clear cell renal cell carcinoma',\n",
       " 'coad': 'Colon adenocarcinoma',\n",
       " 'gbm': 'Glioblastoma multiforme',\n",
       " 'hnscc': 'Head and Neck squamous cell carcinoma',\n",
       " 'lscc': 'Lung squamous cell carcinoma',\n",
       " 'luad': 'Lung adenocarcinoma',\n",
       " 'ov': 'Ovarian serous cystadenocarcinoma',\n",
       " 'pda': 'Pancreatic ductal adenocarcinoma',\n",
       " 'pdac': 'Pancreatic ductal adenocarcinoma',\n",
       " 'ucec': 'Uterine Corpus Endometrial Carcinoma'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cptac.get_cancer_info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now to understand the data types that are present, ######### BLAKE FIGURE IT OUT. call the `cptac.get_cancer_options()` function. You will note that sometimes a datatype has multiple sources. These represent different computational pipelines run by different bioinformatics groups. For example, the transcriptomics data has both `bcm` and `broad` as sources, designating that there are different versions of transcriptomics data create by the groups at Baylor College of Medicine and The Broad. As shown down below, wherever you request a dataframe, you need to specify a source."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Cancer</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Source</th>\n",
       "      <th>Datatype</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">bcm</th>\n",
       "      <th>circular_RNA</th>\n",
       "      <td>[lscc, hnscc, gbm, ucec, pdac, ccrcc, luad]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcriptomics</th>\n",
       "      <td>[coad, lscc, hnscc, gbm, ov, ucec, pdac, brca,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>broad</th>\n",
       "      <th>transcriptomics</th>\n",
       "      <td>[coad, lscc, hnscc, gbm, ov, ucec, pdac, brca,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">harmonized</th>\n",
       "      <th>ancestry_prediction</th>\n",
       "      <td>[all_cancers]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>somatic_mutation</th>\n",
       "      <td>[all_cancers]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mssm</th>\n",
       "      <th>clinical</th>\n",
       "      <td>[all_cancers]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">umich</th>\n",
       "      <th>acetylproteomics</th>\n",
       "      <td>[lscc, gbm, ucec, brca, luad]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phosphoproteomics</th>\n",
       "      <td>[coad, lscc, hnscc, gbm, ucec, pdac, brca, ccr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>proteomics</th>\n",
       "      <td>[coad, lscc, hnscc, gbm, ov, ucec, pdac, brca,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">washu</th>\n",
       "      <th>CNV</th>\n",
       "      <td>[coad, lscc, hnscc, gbm, ov, ucec, pdac, brca,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cibersort</th>\n",
       "      <td>[coad, lscc, hnscc, gbm, ov, ucec, pdac, brca,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hla_typing</th>\n",
       "      <td>[all_cancers]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>somatic_mutation</th>\n",
       "      <td>[coad, lscc, hnscc, gbm, ov, ucec, pdac, luad]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcriptomics</th>\n",
       "      <td>[coad, hnscc, gbm, ov, ucec, pdac, brca, ccrcc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumor_purity</th>\n",
       "      <td>[all_cancers]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>xcell</th>\n",
       "      <td>[coad, lscc, hnscc, gbm, ov, ucec, pdac, brca,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                           Cancer\n",
       "Source     Datatype                                                              \n",
       "bcm        circular_RNA               [lscc, hnscc, gbm, ucec, pdac, ccrcc, luad]\n",
       "           transcriptomics      [coad, lscc, hnscc, gbm, ov, ucec, pdac, brca,...\n",
       "broad      transcriptomics      [coad, lscc, hnscc, gbm, ov, ucec, pdac, brca,...\n",
       "harmonized ancestry_prediction                                      [all_cancers]\n",
       "           somatic_mutation                                         [all_cancers]\n",
       "mssm       clinical                                                 [all_cancers]\n",
       "umich      acetylproteomics                         [lscc, gbm, ucec, brca, luad]\n",
       "           phosphoproteomics    [coad, lscc, hnscc, gbm, ucec, pdac, brca, ccr...\n",
       "           proteomics           [coad, lscc, hnscc, gbm, ov, ucec, pdac, brca,...\n",
       "washu      CNV                  [coad, lscc, hnscc, gbm, ov, ucec, pdac, brca,...\n",
       "           cibersort            [coad, lscc, hnscc, gbm, ov, ucec, pdac, brca,...\n",
       "           hla_typing                                               [all_cancers]\n",
       "           somatic_mutation        [coad, lscc, hnscc, gbm, ov, ucec, pdac, luad]\n",
       "           transcriptomics      [coad, hnscc, gbm, ov, ucec, pdac, brca, ccrcc...\n",
       "           tumor_purity                                             [all_cancers]\n",
       "           xcell                [coad, lscc, hnscc, gbm, ov, ucec, pdac, brca,..."
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cptac.get_source_options()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exploring the data\n",
    "\n",
    "`cptac` allows you to load the dataset into a Python variable, and you can use that variable to access and work with the data. To load a particular dataset into a variable, type the name you want to give the variable, followed by `=`, and then type `cptac.` and the name of the dataset in [UpperCamelCase](https://en.wikipedia.org/wiki/Camel_case) followed by two parentheses, e.g. `cptac.Ucec()` or `cptac.Ccrcc()`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "en = cptac.Ucec()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To see what data is available, use the `en.list_data_sources()` function. This displays the different types of data included in the dataset for this particular cancer type, each stored in a [pandas dataframe](https://pandas.pydata.org/pandas-docs/stable/getting_started/dsintro.html#dataframe)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data type</th>\n",
       "      <th>Available sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CNV</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>acetylproteomics</td>\n",
       "      <td>umich</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ancestry_prediction</td>\n",
       "      <td>harmonized</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cibersort</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>circular_RNA</td>\n",
       "      <td>bcm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>clinical</td>\n",
       "      <td>mssm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>follow-up</td>\n",
       "      <td>mssm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>hla_typing</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>medical_history</td>\n",
       "      <td>mssm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>miRNA</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>phosphoproteomics</td>\n",
       "      <td>umich</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>proteomics</td>\n",
       "      <td>umich</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>somatic_mutation</td>\n",
       "      <td>harmonized, washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>bcm, broad, washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>tumor_purity</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>xcell</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Data type  Available sources\n",
       "0                   CNV              washu\n",
       "1      acetylproteomics              umich\n",
       "2   ancestry_prediction         harmonized\n",
       "3             cibersort              washu\n",
       "4          circular_RNA                bcm\n",
       "5              clinical               mssm\n",
       "6             follow-up               mssm\n",
       "7            hla_typing              washu\n",
       "8       medical_history               mssm\n",
       "9                 miRNA              washu\n",
       "10    phosphoproteomics              umich\n",
       "11           proteomics              umich\n",
       "12     somatic_mutation  harmonized, washu\n",
       "13      transcriptomics  bcm, broad, washu\n",
       "14         tumor_purity              washu\n",
       "15                xcell              washu"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "en.list_data_sources()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Molecular Omics\n",
    "\n",
    "Data can be accessed through the `get_dataframe` function, or through one of several helper \"get\" functions. For example, we can look at the proteomics data by using `en.get_proteomics()`. This returns a [pandas dataframe](https://pandas.pydata.org/pandas-docs/stable/getting_started/dsintro.html#dataframe) containing the proteomic data. Each column in the proteomics dataframe is the quantitiative measurement for a particular protein. Each row in the proteomics dataframe is a sample of either a tumor or non-tumor from a cancer patient."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Samples: ['C3L-00006', 'C3L-00008', 'C3L-00032', 'C3L-00084', 'C3L-00090', 'C3L-00098', 'C3L-00136', 'C3L-00137', 'C3L-00139', 'C3L-00143', 'C3L-00145', 'C3L-00156', 'C3L-00157', 'C3L-00161', 'C3L-00356', 'C3L-00358', 'C3L-00361', 'C3L-00362', 'C3L-00413', 'C3L-00449']\n",
      "Proteins: [('ARF5', 'ENSP00000000233.5'), ('M6PR', 'ENSP00000000412.3'), ('ESRRA', 'ENSP00000000442.6'), ('FKBP4', 'ENSP00000001008.4'), ('NDUFAF7', 'ENSP00000002125.4'), ('FUCA2', 'ENSP00000002165.5'), ('DBNDD1', 'ENSP00000002501.6'), ('SEMA3F', 'ENSP00000002829.3'), ('CFTR', 'ENSP00000003084.6'), ('CYP51A1', 'ENSP00000003100.8'), ('USP28', 'ENSP00000003302.4'), ('SLC7A2', 'ENSP00000004531.10'), ('HSPB6', 'ENSP00000004982.3'), ('PDK4', 'ENSP00000005178.5'), ('USH1C', 'ENSP00000005226.7'), ('RALA', 'ENSP00000005257.2'), ('BAIAP2L1', 'ENSP00000005260.8'), ('TMEM132A', 'ENSP00000005286.4'), ('DVL2', 'ENSP00000005340.4'), ('RPAP3', 'ENSP00000005386.3')]\n"
     ]
    }
   ],
   "source": [
    "# These two methods of getting proteomics data are functionally equivalent:\n",
    "# get_dataframe needs at least two arguments: datatype, and source\n",
    "proteomics = en.get_dataframe('proteomics', 'umich')\n",
    "# there is also a get function for each datatype, this is the same as get_dataframe \n",
    "# but does not need the datatype argument. This function call below is equivalent to \n",
    "# the one above\n",
    "proteomics = en.get_proteomics('umich')\n",
    "\n",
    "samples = proteomics.index\n",
    "proteins = proteomics.columns\n",
    "print(\"Samples:\",samples[0:20].tolist()) #the first twenty samples\n",
    "print(\"Proteins:\",proteins[0:20].tolist()) #the first twenty proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dataframe values\n",
    "\n",
    "Values in the dataframe are protein abundance values. Values that read \"NaN\" mean that particular sample from that patient had no data for that particular protein. In CPTAC proteomics data, a TMT-reference channel strategy was used. A detailed description of this strategy can be found at [Nature Protocols](https://www.nature.com/articles/s41596-018-0006-9) and also at [PubMed Central](https://www.ncbi.nlm.nih.gov/pubmed/?term=29988108). This strategy ratios each sample's abundance to a pooled reference. The ratio is then log transformed. Therefore positive values indicate a measurement higher than the pooled reference; negative values are lower than the pooled reference."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th>ARF5</th>\n",
       "      <th>M6PR</th>\n",
       "      <th>ESRRA</th>\n",
       "      <th>FKBP4</th>\n",
       "      <th>NDUFAF7</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>DBNDD1</th>\n",
       "      <th>SEMA3F</th>\n",
       "      <th>CFTR</th>\n",
       "      <th>CYP51A1</th>\n",
       "      <th>...</th>\n",
       "      <th>SCRIB</th>\n",
       "      <th>WIZ</th>\n",
       "      <th>BPIFB4</th>\n",
       "      <th>LDB1</th>\n",
       "      <th>WIZ</th>\n",
       "      <th>TSGA10</th>\n",
       "      <th>RFX7</th>\n",
       "      <th>SWSAP1</th>\n",
       "      <th>MSANTD2</th>\n",
       "      <th>SVIL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Database_ID</th>\n",
       "      <th>ENSP00000000233.5</th>\n",
       "      <th>ENSP00000000412.3</th>\n",
       "      <th>ENSP00000000442.6</th>\n",
       "      <th>ENSP00000001008.4</th>\n",
       "      <th>ENSP00000002125.4</th>\n",
       "      <th>ENSP00000002165.5</th>\n",
       "      <th>ENSP00000002501.6</th>\n",
       "      <th>ENSP00000002829.3</th>\n",
       "      <th>ENSP00000003084.6</th>\n",
       "      <th>ENSP00000003100.8</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSP00000501177.1</th>\n",
       "      <th>ENSP00000501256.3</th>\n",
       "      <th>ENSP00000501266.1</th>\n",
       "      <th>ENSP00000501277.1</th>\n",
       "      <th>ENSP00000501300.1</th>\n",
       "      <th>ENSP00000501312.1</th>\n",
       "      <th>ENSP00000501317.1</th>\n",
       "      <th>ENSP00000501355.1</th>\n",
       "      <th>ENSP00000501466.1</th>\n",
       "      <th>ENSP00000501521.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>-0.056513</td>\n",
       "      <td>0.016557</td>\n",
       "      <td>0.002569</td>\n",
       "      <td>0.389819</td>\n",
       "      <td>0.603610</td>\n",
       "      <td>-0.332543</td>\n",
       "      <td>-0.790426</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.822732</td>\n",
       "      <td>0.039134</td>\n",
       "      <td>...</td>\n",
       "      <td>0.161720</td>\n",
       "      <td>-0.884807</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.268247</td>\n",
       "      <td>0.125392</td>\n",
       "      <td>-0.880833</td>\n",
       "      <td>0.108554</td>\n",
       "      <td>0.107413</td>\n",
       "      <td>-0.085833</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00008</th>\n",
       "      <td>0.549959</td>\n",
       "      <td>-0.206129</td>\n",
       "      <td>0.905784</td>\n",
       "      <td>-0.303631</td>\n",
       "      <td>0.018767</td>\n",
       "      <td>0.503513</td>\n",
       "      <td>0.950955</td>\n",
       "      <td>0.080142</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.063213</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.054284</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.106450</td>\n",
       "      <td>0.380557</td>\n",
       "      <td>-0.756099</td>\n",
       "      <td>0.264611</td>\n",
       "      <td>0.044423</td>\n",
       "      <td>-0.248319</td>\n",
       "      <td>-1.206596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00032</th>\n",
       "      <td>0.088681</td>\n",
       "      <td>-0.154447</td>\n",
       "      <td>-0.190515</td>\n",
       "      <td>0.170753</td>\n",
       "      <td>0.196356</td>\n",
       "      <td>0.544194</td>\n",
       "      <td>-0.179078</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.377405</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.086905</td>\n",
       "      <td>0.055991</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.021986</td>\n",
       "      <td>-0.229645</td>\n",
       "      <td>1.923986</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.176694</td>\n",
       "      <td>-0.332384</td>\n",
       "      <td>-1.330653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00084</th>\n",
       "      <td>-0.846555</td>\n",
       "      <td>0.027740</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.178700</td>\n",
       "      <td>0.264054</td>\n",
       "      <td>-0.183548</td>\n",
       "      <td>0.077215</td>\n",
       "      <td>-0.247164</td>\n",
       "      <td>0.152277</td>\n",
       "      <td>-0.279549</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.125796</td>\n",
       "      <td>0.944212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.917409</td>\n",
       "      <td>0.026862</td>\n",
       "      <td>-0.885976</td>\n",
       "      <td>-0.006510</td>\n",
       "      <td>-0.014162</td>\n",
       "      <td>0.365158</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00090</th>\n",
       "      <td>0.539019</td>\n",
       "      <td>0.956619</td>\n",
       "      <td>-0.039516</td>\n",
       "      <td>0.323656</td>\n",
       "      <td>0.064605</td>\n",
       "      <td>0.173433</td>\n",
       "      <td>-0.524325</td>\n",
       "      <td>-0.038590</td>\n",
       "      <td>-0.311486</td>\n",
       "      <td>0.309905</td>\n",
       "      <td>...</td>\n",
       "      <td>0.853362</td>\n",
       "      <td>-0.716947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.286277</td>\n",
       "      <td>-0.046076</td>\n",
       "      <td>0.089645</td>\n",
       "      <td>-0.444506</td>\n",
       "      <td>-0.072531</td>\n",
       "      <td>-0.463495</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 12662 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name                     ARF5              M6PR             ESRRA  \\\n",
       "Database_ID ENSP00000000233.5 ENSP00000000412.3 ENSP00000000442.6   \n",
       "Patient_ID                                                          \n",
       "C3L-00006           -0.056513          0.016557          0.002569   \n",
       "C3L-00008            0.549959         -0.206129          0.905784   \n",
       "C3L-00032            0.088681         -0.154447         -0.190515   \n",
       "C3L-00084           -0.846555          0.027740               NaN   \n",
       "C3L-00090            0.539019          0.956619         -0.039516   \n",
       "\n",
       "Name                    FKBP4           NDUFAF7             FUCA2  \\\n",
       "Database_ID ENSP00000001008.4 ENSP00000002125.4 ENSP00000002165.5   \n",
       "Patient_ID                                                          \n",
       "C3L-00006            0.389819          0.603610         -0.332543   \n",
       "C3L-00008           -0.303631          0.018767          0.503513   \n",
       "C3L-00032            0.170753          0.196356          0.544194   \n",
       "C3L-00084            0.178700          0.264054         -0.183548   \n",
       "C3L-00090            0.323656          0.064605          0.173433   \n",
       "\n",
       "Name                   DBNDD1            SEMA3F              CFTR  \\\n",
       "Database_ID ENSP00000002501.6 ENSP00000002829.3 ENSP00000003084.6   \n",
       "Patient_ID                                                          \n",
       "C3L-00006           -0.790426               NaN          0.822732   \n",
       "C3L-00008            0.950955          0.080142               NaN   \n",
       "C3L-00032           -0.179078               NaN               NaN   \n",
       "C3L-00084            0.077215         -0.247164          0.152277   \n",
       "C3L-00090           -0.524325         -0.038590         -0.311486   \n",
       "\n",
       "Name                  CYP51A1  ...             SCRIB               WIZ  \\\n",
       "Database_ID ENSP00000003100.8  ... ENSP00000501177.1 ENSP00000501256.3   \n",
       "Patient_ID                     ...                                       \n",
       "C3L-00006            0.039134  ...          0.161720         -0.884807   \n",
       "C3L-00008           -0.063213  ...               NaN          0.054284   \n",
       "C3L-00032            0.377405  ...         -1.086905          0.055991   \n",
       "C3L-00084           -0.279549  ...         -0.125796          0.944212   \n",
       "C3L-00090            0.309905  ...          0.853362         -0.716947   \n",
       "\n",
       "Name                   BPIFB4              LDB1               WIZ  \\\n",
       "Database_ID ENSP00000501266.1 ENSP00000501277.1 ENSP00000501300.1   \n",
       "Patient_ID                                                          \n",
       "C3L-00006                 NaN          0.268247          0.125392   \n",
       "C3L-00008                 NaN         -0.106450          0.380557   \n",
       "C3L-00032                 NaN         -0.021986         -0.229645   \n",
       "C3L-00084                 NaN          0.917409          0.026862   \n",
       "C3L-00090                 NaN         -0.286277         -0.046076   \n",
       "\n",
       "Name                   TSGA10              RFX7            SWSAP1  \\\n",
       "Database_ID ENSP00000501312.1 ENSP00000501317.1 ENSP00000501355.1   \n",
       "Patient_ID                                                          \n",
       "C3L-00006           -0.880833          0.108554          0.107413   \n",
       "C3L-00008           -0.756099          0.264611          0.044423   \n",
       "C3L-00032            1.923986               NaN         -0.176694   \n",
       "C3L-00084           -0.885976         -0.006510         -0.014162   \n",
       "C3L-00090            0.089645         -0.444506         -0.072531   \n",
       "\n",
       "Name                  MSANTD2              SVIL  \n",
       "Database_ID ENSP00000501466.1 ENSP00000501521.1  \n",
       "Patient_ID                                       \n",
       "C3L-00006           -0.085833               NaN  \n",
       "C3L-00008           -0.248319         -1.206596  \n",
       "C3L-00032           -0.332384         -1.330653  \n",
       "C3L-00084            0.365158               NaN  \n",
       "C3L-00090           -0.463495               NaN  \n",
       "\n",
       "[5 rows x 12662 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteomics.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As seen in `en.list_data_sources()`, other omics data are also available (e.g. transcriptomics, copy number variation, phoshoproteomics).\n",
    "\n",
    "The transcriptomics looks almost identical to the proteomics data, available in a pandas dataframe with the same convention. Each set of samples is consistent, meaning samples found in the endometrial proteomics data will be the same samples in all other endometrial dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2ML1-AS1</th>\n",
       "      <th>A2ML1-AS2</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>...</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11AP1</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZYXP1</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>hsa-mir-1253</th>\n",
       "      <th>hsa-mir-423</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Database_ID</th>\n",
       "      <th>ENSG00000121410.12</th>\n",
       "      <th>ENSG00000268895.6</th>\n",
       "      <th>ENSG00000148584.15</th>\n",
       "      <th>ENSG00000175899.15</th>\n",
       "      <th>ENSG00000245105.4</th>\n",
       "      <th>ENSG00000166535.20</th>\n",
       "      <th>ENSG00000256661.1</th>\n",
       "      <th>ENSG00000256904.1</th>\n",
       "      <th>ENSG00000256069.7</th>\n",
       "      <th>ENSG00000184389.9</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000198455.4</th>\n",
       "      <th>ENSG00000070476.15</th>\n",
       "      <th>ENSG00000203995.10</th>\n",
       "      <th>ENSG00000232242.2</th>\n",
       "      <th>ENSG00000162378.13</th>\n",
       "      <th>ENSG00000159840.16</th>\n",
       "      <th>ENSG00000274572.1</th>\n",
       "      <th>ENSG00000074755.15</th>\n",
       "      <th>ENSG00000272920.1</th>\n",
       "      <th>ENSG00000266919.3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>2.54</td>\n",
       "      <td>5.11</td>\n",
       "      <td>3.60</td>\n",
       "      <td>13.75</td>\n",
       "      <td>6.45</td>\n",
       "      <td>7.08</td>\n",
       "      <td>1.80</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.60</td>\n",
       "      <td>1.16</td>\n",
       "      <td>...</td>\n",
       "      <td>10.17</td>\n",
       "      <td>10.61</td>\n",
       "      <td>5.54</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.85</td>\n",
       "      <td>10.60</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.87</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00008</th>\n",
       "      <td>4.40</td>\n",
       "      <td>4.63</td>\n",
       "      <td>5.49</td>\n",
       "      <td>13.89</td>\n",
       "      <td>6.61</td>\n",
       "      <td>6.97</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.74</td>\n",
       "      <td>3.25</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>9.79</td>\n",
       "      <td>10.48</td>\n",
       "      <td>7.79</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.28</td>\n",
       "      <td>11.28</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.93</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00032</th>\n",
       "      <td>4.83</td>\n",
       "      <td>7.26</td>\n",
       "      <td>3.73</td>\n",
       "      <td>14.48</td>\n",
       "      <td>6.91</td>\n",
       "      <td>9.56</td>\n",
       "      <td>0.98</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.26</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>9.43</td>\n",
       "      <td>9.97</td>\n",
       "      <td>6.48</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.72</td>\n",
       "      <td>10.37</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.70</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00084</th>\n",
       "      <td>4.73</td>\n",
       "      <td>6.01</td>\n",
       "      <td>5.37</td>\n",
       "      <td>15.17</td>\n",
       "      <td>7.93</td>\n",
       "      <td>3.86</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.73</td>\n",
       "      <td>1.15</td>\n",
       "      <td>...</td>\n",
       "      <td>9.23</td>\n",
       "      <td>10.37</td>\n",
       "      <td>7.47</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.86</td>\n",
       "      <td>10.13</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.19</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00090</th>\n",
       "      <td>4.14</td>\n",
       "      <td>6.24</td>\n",
       "      <td>5.69</td>\n",
       "      <td>13.87</td>\n",
       "      <td>6.79</td>\n",
       "      <td>4.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.23</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>9.69</td>\n",
       "      <td>9.64</td>\n",
       "      <td>7.60</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.98</td>\n",
       "      <td>10.31</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.45</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 59286 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name                      A1BG          A1BG-AS1               A1CF  \\\n",
       "Database_ID ENSG00000121410.12 ENSG00000268895.6 ENSG00000148584.15   \n",
       "Patient_ID                                                            \n",
       "C3L-00006                 2.54              5.11               3.60   \n",
       "C3L-00008                 4.40              4.63               5.49   \n",
       "C3L-00032                 4.83              7.26               3.73   \n",
       "C3L-00084                 4.73              6.01               5.37   \n",
       "C3L-00090                 4.14              6.24               5.69   \n",
       "\n",
       "Name                       A2M           A2M-AS1              A2ML1  \\\n",
       "Database_ID ENSG00000175899.15 ENSG00000245105.4 ENSG00000166535.20   \n",
       "Patient_ID                                                            \n",
       "C3L-00006                13.75              6.45               7.08   \n",
       "C3L-00008                13.89              6.61               6.97   \n",
       "C3L-00032                14.48              6.91               9.56   \n",
       "C3L-00084                15.17              7.93               3.86   \n",
       "C3L-00090                13.87              6.79               4.32   \n",
       "\n",
       "Name                A2ML1-AS1         A2ML1-AS2             A2MP1  \\\n",
       "Database_ID ENSG00000256661.1 ENSG00000256904.1 ENSG00000256069.7   \n",
       "Patient_ID                                                          \n",
       "C3L-00006                1.80              0.00              2.60   \n",
       "C3L-00008                0.00              2.74              3.25   \n",
       "C3L-00032                0.98              0.00              3.26   \n",
       "C3L-00084                0.00              0.00              3.73   \n",
       "C3L-00090                0.00              0.00              3.23   \n",
       "\n",
       "Name                  A3GALT2  ...              ZXDB               ZXDC  \\\n",
       "Database_ID ENSG00000184389.9  ... ENSG00000198455.4 ENSG00000070476.15   \n",
       "Patient_ID                     ...                                        \n",
       "C3L-00006                1.16  ...             10.17              10.61   \n",
       "C3L-00008                0.00  ...              9.79              10.48   \n",
       "C3L-00032                0.00  ...              9.43               9.97   \n",
       "C3L-00084                1.15  ...              9.23              10.37   \n",
       "C3L-00090                0.00  ...              9.69               9.64   \n",
       "\n",
       "Name                    ZYG11A          ZYG11AP1             ZYG11B  \\\n",
       "Database_ID ENSG00000203995.10 ENSG00000232242.2 ENSG00000162378.13   \n",
       "Patient_ID                                                            \n",
       "C3L-00006                 5.54               0.0              11.85   \n",
       "C3L-00008                 7.79               0.0              12.28   \n",
       "C3L-00032                 6.48               0.0              11.72   \n",
       "C3L-00084                 7.47               0.0              11.86   \n",
       "C3L-00090                 7.60               0.0              11.98   \n",
       "\n",
       "Name                       ZYX             ZYXP1              ZZEF1  \\\n",
       "Database_ID ENSG00000159840.16 ENSG00000274572.1 ENSG00000074755.15   \n",
       "Patient_ID                                                            \n",
       "C3L-00006                10.60               0.0              11.87   \n",
       "C3L-00008                11.28               0.0              11.93   \n",
       "C3L-00032                10.37               0.0              11.70   \n",
       "C3L-00084                10.13               0.0              11.19   \n",
       "C3L-00090                10.31               0.0              11.45   \n",
       "\n",
       "Name             hsa-mir-1253       hsa-mir-423  \n",
       "Database_ID ENSG00000272920.1 ENSG00000266919.3  \n",
       "Patient_ID                                       \n",
       "C3L-00006                 0.0               0.0  \n",
       "C3L-00008                 0.0               0.0  \n",
       "C3L-00032                 0.0               0.0  \n",
       "C3L-00084                 0.0               0.0  \n",
       "C3L-00090                 0.0               0.0  \n",
       "\n",
       "[5 rows x 59286 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcriptomics = en.get_transcriptomics('bcm')\n",
    "transcriptomics.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clinical Data\n",
    "\n",
    "The clinical dataframe lists clinical information for the patient associated with each sample (e.g. age, race, diabetes status, tumor size). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>tumor_code</th>\n",
       "      <th>discovery_study</th>\n",
       "      <th>type_of_analyzed_samples</th>\n",
       "      <th>confirmatory_study</th>\n",
       "      <th>type_of_analyzed_samples</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>ethnicity_race_ancestry_identified</th>\n",
       "      <th>...</th>\n",
       "      <th>additional_treatment_pharmaceutical_therapy_for_new_tumor</th>\n",
       "      <th>additional_treatment_immuno_for_new_tumor</th>\n",
       "      <th>number_of_days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor_event_loco-regional</th>\n",
       "      <th>number_of_days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor_event_metastasis</th>\n",
       "      <th>Recurrence-free survival, days</th>\n",
       "      <th>Recurrence-free survival from collection, days</th>\n",
       "      <th>Recurrence status (1, yes; 0, no)</th>\n",
       "      <th>Overall survival, days</th>\n",
       "      <th>Overall survival from collection, days</th>\n",
       "      <th>Survival status (1, dead; 0, alive)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor_and_Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>64</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>737.0</td>\n",
       "      <td>737.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00008</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>58</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>898.0</td>\n",
       "      <td>898.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00032</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1710.0</td>\n",
       "      <td>1710.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00084</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>335.0</td>\n",
       "      <td>335.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00090</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1281.0</td>\n",
       "      <td>1287.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 124 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       tumor_code discovery_study type_of_analyzed_samples  \\\n",
       "Patient_ID                                                       \n",
       "C3L-00006        UCEC             Yes         Tumor_and_Normal   \n",
       "C3L-00008        UCEC             Yes                    Tumor   \n",
       "C3L-00032        UCEC             Yes                    Tumor   \n",
       "C3L-00084        UCEC             Yes                    Tumor   \n",
       "C3L-00090        UCEC             Yes                    Tumor   \n",
       "\n",
       "Name       confirmatory_study type_of_analyzed_samples age     sex   race  \\\n",
       "Patient_ID                                                                  \n",
       "C3L-00006                 NaN                      NaN  64  Female  White   \n",
       "C3L-00008                 NaN                      NaN  58  Female  White   \n",
       "C3L-00032                 NaN                      NaN  50  Female  White   \n",
       "C3L-00084                 NaN                      NaN  74  Female  White   \n",
       "C3L-00090                 NaN                      NaN  75  Female  White   \n",
       "\n",
       "Name                     ethnicity ethnicity_race_ancestry_identified  ...  \\\n",
       "Patient_ID                                                             ...   \n",
       "C3L-00006   Not Hispanic or Latino                              White  ...   \n",
       "C3L-00008   Not Hispanic or Latino                              White  ...   \n",
       "C3L-00032   Not Hispanic or Latino                              White  ...   \n",
       "C3L-00084   Not Hispanic or Latino                              White  ...   \n",
       "C3L-00090   Not Hispanic or Latino                              White  ...   \n",
       "\n",
       "Name       additional_treatment_pharmaceutical_therapy_for_new_tumor  \\\n",
       "Patient_ID                                                             \n",
       "C3L-00006                                                 NaN          \n",
       "C3L-00008                                                 NaN          \n",
       "C3L-00032                                                 NaN          \n",
       "C3L-00084                                                 NaN          \n",
       "C3L-00090                                                 Yes          \n",
       "\n",
       "Name       additional_treatment_immuno_for_new_tumor  \\\n",
       "Patient_ID                                             \n",
       "C3L-00006                                        NaN   \n",
       "C3L-00008                                        NaN   \n",
       "C3L-00032                                        NaN   \n",
       "C3L-00084                                        NaN   \n",
       "C3L-00090                                         No   \n",
       "\n",
       "Name       number_of_days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor_event_loco-regional  \\\n",
       "Patient_ID                                                                                                                            \n",
       "C3L-00006                                                 NaN                                                                         \n",
       "C3L-00008                                                 NaN                                                                         \n",
       "C3L-00032                                                 NaN                                                                         \n",
       "C3L-00084                                                 NaN                                                                         \n",
       "C3L-00090                                                 NaN                                                                         \n",
       "\n",
       "Name       number_of_days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor_event_metastasis  \\\n",
       "Patient_ID                                                                                                                         \n",
       "C3L-00006                                                 NaN                                                                      \n",
       "C3L-00008                                                 NaN                                                                      \n",
       "C3L-00032                                                 NaN                                                                      \n",
       "C3L-00084                                                 NaN                                                                      \n",
       "C3L-00090                                                 NaN                                                                      \n",
       "\n",
       "Name       Recurrence-free survival, days  \\\n",
       "Patient_ID                                  \n",
       "C3L-00006                             NaN   \n",
       "C3L-00008                             NaN   \n",
       "C3L-00032                             NaN   \n",
       "C3L-00084                             NaN   \n",
       "C3L-00090                            50.0   \n",
       "\n",
       "Name       Recurrence-free survival from collection, days  \\\n",
       "Patient_ID                                                  \n",
       "C3L-00006                                             NaN   \n",
       "C3L-00008                                             NaN   \n",
       "C3L-00032                                             NaN   \n",
       "C3L-00084                                             NaN   \n",
       "C3L-00090                                            56.0   \n",
       "\n",
       "Name       Recurrence status (1, yes; 0, no) Overall survival, days  \\\n",
       "Patient_ID                                                            \n",
       "C3L-00006                                  0                  737.0   \n",
       "C3L-00008                                  0                  898.0   \n",
       "C3L-00032                                  0                 1710.0   \n",
       "C3L-00084                                  0                  335.0   \n",
       "C3L-00090                                  1                 1281.0   \n",
       "\n",
       "Name       Overall survival from collection, days  \\\n",
       "Patient_ID                                          \n",
       "C3L-00006                                   737.0   \n",
       "C3L-00008                                   898.0   \n",
       "C3L-00032                                  1710.0   \n",
       "C3L-00084                                   335.0   \n",
       "C3L-00090                                  1287.0   \n",
       "\n",
       "Name       Survival status (1, dead; 0, alive)  \n",
       "Patient_ID                                      \n",
       "C3L-00006                                  0.0  \n",
       "C3L-00008                                  0.0  \n",
       "C3L-00032                                  0.0  \n",
       "C3L-00084                                  0.0  \n",
       "C3L-00090                                  1.0  \n",
       "\n",
       "[5 rows x 124 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical = en.get_clinical('mssm')\n",
    "clinical.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition to donating a tumor sample, some patients also had a normal sample taken for control and comparison. We can identify these samples by looking for samples marked \"Tumor_and_Normal\" in the \"type_of_analyzed_samples\" column. For example, patient C3L-00006 provided both a tumor sample and a normal sample. Since the data in these cases is for both tumor and normal samples, where a column would have identical values for both samples from the same patient (e.g., patient age and gender), the information is recorded only for the tumor sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>tumor_code</th>\n",
       "      <th>discovery_study</th>\n",
       "      <th>type_of_analyzed_samples</th>\n",
       "      <th>confirmatory_study</th>\n",
       "      <th>type_of_analyzed_samples</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>ethnicity_race_ancestry_identified</th>\n",
       "      <th>...</th>\n",
       "      <th>additional_treatment_pharmaceutical_therapy_for_new_tumor</th>\n",
       "      <th>additional_treatment_immuno_for_new_tumor</th>\n",
       "      <th>number_of_days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor_event_loco-regional</th>\n",
       "      <th>number_of_days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor_event_metastasis</th>\n",
       "      <th>Recurrence-free survival, days</th>\n",
       "      <th>Recurrence-free survival from collection, days</th>\n",
       "      <th>Recurrence status (1, yes; 0, no)</th>\n",
       "      <th>Overall survival, days</th>\n",
       "      <th>Overall survival from collection, days</th>\n",
       "      <th>Survival status (1, dead; 0, alive)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor_and_Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>64</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>White</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>737.0</td>\n",
       "      <td>737.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00361</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor_and_Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>64</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Hispanic or Latino</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1707.0</td>\n",
       "      <td>1707.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-01246</th>\n",
       "      <td>UCEC</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Tumor_and_Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>62</td>\n",
       "      <td>Female</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Medical record does not state.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>1538.0</td>\n",
       "      <td>1524.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows Ã— 124 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       tumor_code discovery_study type_of_analyzed_samples  \\\n",
       "Patient_ID                                                       \n",
       "C3L-00006        UCEC             Yes         Tumor_and_Normal   \n",
       "C3L-00361        UCEC             Yes         Tumor_and_Normal   \n",
       "C3L-01246        UCEC             Yes         Tumor_and_Normal   \n",
       "\n",
       "Name       confirmatory_study type_of_analyzed_samples age     sex   race  \\\n",
       "Patient_ID                                                                  \n",
       "C3L-00006                 NaN                      NaN  64  Female  White   \n",
       "C3L-00361                 NaN                      NaN  64  Female  White   \n",
       "C3L-01246                 NaN                      NaN  62  Female  White   \n",
       "\n",
       "Name                     ethnicity ethnicity_race_ancestry_identified  ...  \\\n",
       "Patient_ID                                                             ...   \n",
       "C3L-00006   Not Hispanic or Latino                              White  ...   \n",
       "C3L-00361   Not Hispanic or Latino                              WHITE  ...   \n",
       "C3L-01246                  Unknown     Medical record does not state.  ...   \n",
       "\n",
       "Name       additional_treatment_pharmaceutical_therapy_for_new_tumor  \\\n",
       "Patient_ID                                                             \n",
       "C3L-00006                                                 NaN          \n",
       "C3L-00361                                                 NaN          \n",
       "C3L-01246                                                 NaN          \n",
       "\n",
       "Name       additional_treatment_immuno_for_new_tumor  \\\n",
       "Patient_ID                                             \n",
       "C3L-00006                                        NaN   \n",
       "C3L-00361                                        NaN   \n",
       "C3L-01246                                        NaN   \n",
       "\n",
       "Name       number_of_days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor_event_loco-regional  \\\n",
       "Patient_ID                                                                                                                            \n",
       "C3L-00006                                                 NaN                                                                         \n",
       "C3L-00361                                                 NaN                                                                         \n",
       "C3L-01246                                                 NaN                                                                         \n",
       "\n",
       "Name       number_of_days_from_date_of_initial_pathologic_diagnosis_to_date_of_additional_surgery_for_new_tumor_event_metastasis  \\\n",
       "Patient_ID                                                                                                                         \n",
       "C3L-00006                                                 NaN                                                                      \n",
       "C3L-00361                                                 NaN                                                                      \n",
       "C3L-01246                                                 NaN                                                                      \n",
       "\n",
       "Name       Recurrence-free survival, days  \\\n",
       "Patient_ID                                  \n",
       "C3L-00006                             NaN   \n",
       "C3L-00361                             NaN   \n",
       "C3L-01246                             NaN   \n",
       "\n",
       "Name       Recurrence-free survival from collection, days  \\\n",
       "Patient_ID                                                  \n",
       "C3L-00006                                             NaN   \n",
       "C3L-00361                                             NaN   \n",
       "C3L-01246                                             NaN   \n",
       "\n",
       "Name       Recurrence status (1, yes; 0, no) Overall survival, days  \\\n",
       "Patient_ID                                                            \n",
       "C3L-00006                                  0                  737.0   \n",
       "C3L-00361                                  0                 1707.0   \n",
       "C3L-01246                                  0                 1538.0   \n",
       "\n",
       "Name       Overall survival from collection, days  \\\n",
       "Patient_ID                                          \n",
       "C3L-00006                                   737.0   \n",
       "C3L-00361                                  1707.0   \n",
       "C3L-01246                                  1524.0   \n",
       "\n",
       "Name       Survival status (1, dead; 0, alive)  \n",
       "Patient_ID                                      \n",
       "C3L-00006                                  0.0  \n",
       "C3L-00361                                  0.0  \n",
       "C3L-01246                                  0.0  \n",
       "\n",
       "[3 rows x 124 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical.loc[[\"C3L-00006\",\"C3L-00361\",\"C3L-01246\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutation data\n",
    "\n",
    "Each cancer dataset contains mutation data for the cohort. The data consists of all somatic mutations found for each sample (meaning there will be many lines for each sample). Each row lists the specific gene that was mutated, the type of mutation, and the location of the mutation. This data is a direct import of a MAF file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading PanCan_Union_Maf_Broad_WashU_v1.1.maf.gz: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 138M/138M [02:55<00:00, 785kB/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Mutation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>...</th>\n",
       "      <th>HGNC_UniProt_ID(supplied_by_UniProt)</th>\n",
       "      <th>HGNC_Ensembl_ID(supplied_by_Ensembl)</th>\n",
       "      <th>HGNC_UCSC_ID(supplied_by_UCSC)</th>\n",
       "      <th>Oreganno_Build</th>\n",
       "      <th>Simple_Uniprot_alt_uniprot_accessions</th>\n",
       "      <th>dbSNP_TOPMED</th>\n",
       "      <th>HGNC_Entrez_Gene_ID(supplied_by_NCBI)</th>\n",
       "      <th>COHORT</th>\n",
       "      <th>getz</th>\n",
       "      <th>washu</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>TEKT5</td>\n",
       "      <td>Silent</td>\n",
       "      <td>p.C178C</td>\n",
       "      <td>146279.0</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr16</td>\n",
       "      <td>10694340</td>\n",
       "      <td>10694340</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>Q96M29</td>\n",
       "      <td>ENSG00000153060</td>\n",
       "      <td>uc002czz.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1L3Z3</td>\n",
       "      <td>0.99969737512742099,0.00030262487257900</td>\n",
       "      <td>146279.0</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>AK9</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.A1819V</td>\n",
       "      <td>221264.0</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr6</td>\n",
       "      <td>109494058</td>\n",
       "      <td>109494058</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>Q5TCS8</td>\n",
       "      <td>ENSG00000155085</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A6NL75|B2RDJ0|B6ZDM7|Q3MIS4|Q5I0W8|Q6ZNF1|Q6ZV...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>221264.0</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>FYN</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.T15M</td>\n",
       "      <td>2534.0</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr6</td>\n",
       "      <td>111720008</td>\n",
       "      <td>111720008</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>P06241</td>\n",
       "      <td>ENSG00000010810</td>\n",
       "      <td>uc003pvk.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B5BU57|E1P557|H0UI48|Q16248|Q5R3A6|Q5R3A7|Q8N5D7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2534.0</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>LAMA4</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.T206I</td>\n",
       "      <td>3910.0</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr6</td>\n",
       "      <td>112191737</td>\n",
       "      <td>112191737</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>Q16363</td>\n",
       "      <td>ENSG00000112769</td>\n",
       "      <td>uc003pvv.4</td>\n",
       "      <td>hg38</td>\n",
       "      <td>Q14731|Q14735|Q15335|Q4LE44|Q5SZG8|Q9BTB8|Q9UE...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3910.0</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>MARCKS</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.A129V</td>\n",
       "      <td>4082.0</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr6</td>\n",
       "      <td>113859966</td>\n",
       "      <td>113859966</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>P29966</td>\n",
       "      <td>ENSG00000277443</td>\n",
       "      <td>uc032xir.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E1P560|Q2LA83|Q5TDB7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4082.0</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 131 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name          Gene           Mutation  Location  Entrez_Gene_Id NCBI_Build  \\\n",
       "Patient_ID                                                                   \n",
       "C3L-00006    TEKT5             Silent   p.C178C        146279.0       hg38   \n",
       "C3L-00006      AK9  Missense_Mutation  p.A1819V        221264.0       hg38   \n",
       "C3L-00006      FYN  Missense_Mutation    p.T15M          2534.0       hg38   \n",
       "C3L-00006    LAMA4  Missense_Mutation   p.T206I          3910.0       hg38   \n",
       "C3L-00006   MARCKS  Missense_Mutation   p.A129V          4082.0       hg38   \n",
       "\n",
       "Name       Chromosome  Start_Position  End_Position Strand Variant_Type  ...  \\\n",
       "Patient_ID                                                               ...   \n",
       "C3L-00006       chr16        10694340      10694340      +          SNP  ...   \n",
       "C3L-00006        chr6       109494058     109494058      +          SNP  ...   \n",
       "C3L-00006        chr6       111720008     111720008      +          SNP  ...   \n",
       "C3L-00006        chr6       112191737     112191737      +          SNP  ...   \n",
       "C3L-00006        chr6       113859966     113859966      +          SNP  ...   \n",
       "\n",
       "Name       HGNC_UniProt_ID(supplied_by_UniProt)  \\\n",
       "Patient_ID                                        \n",
       "C3L-00006                                Q96M29   \n",
       "C3L-00006                                Q5TCS8   \n",
       "C3L-00006                                P06241   \n",
       "C3L-00006                                Q16363   \n",
       "C3L-00006                                P29966   \n",
       "\n",
       "Name       HGNC_Ensembl_ID(supplied_by_Ensembl)  \\\n",
       "Patient_ID                                        \n",
       "C3L-00006                       ENSG00000153060   \n",
       "C3L-00006                       ENSG00000155085   \n",
       "C3L-00006                       ENSG00000010810   \n",
       "C3L-00006                       ENSG00000112769   \n",
       "C3L-00006                       ENSG00000277443   \n",
       "\n",
       "Name       HGNC_UCSC_ID(supplied_by_UCSC) Oreganno_Build  \\\n",
       "Patient_ID                                                 \n",
       "C3L-00006                      uc002czz.2            NaN   \n",
       "C3L-00006                             NaN            NaN   \n",
       "C3L-00006                      uc003pvk.3            NaN   \n",
       "C3L-00006                      uc003pvv.4           hg38   \n",
       "C3L-00006                      uc032xir.2            NaN   \n",
       "\n",
       "Name                    Simple_Uniprot_alt_uniprot_accessions  \\\n",
       "Patient_ID                                                      \n",
       "C3L-00006                                              A1L3Z3   \n",
       "C3L-00006   A6NL75|B2RDJ0|B6ZDM7|Q3MIS4|Q5I0W8|Q6ZNF1|Q6ZV...   \n",
       "C3L-00006    B5BU57|E1P557|H0UI48|Q16248|Q5R3A6|Q5R3A7|Q8N5D7   \n",
       "C3L-00006   Q14731|Q14735|Q15335|Q4LE44|Q5SZG8|Q9BTB8|Q9UE...   \n",
       "C3L-00006                                E1P560|Q2LA83|Q5TDB7   \n",
       "\n",
       "Name                                   dbSNP_TOPMED  \\\n",
       "Patient_ID                                            \n",
       "C3L-00006   0.99969737512742099,0.00030262487257900   \n",
       "C3L-00006                                       NaN   \n",
       "C3L-00006                                       NaN   \n",
       "C3L-00006                                       NaN   \n",
       "C3L-00006                                       NaN   \n",
       "\n",
       "Name        HGNC_Entrez_Gene_ID(supplied_by_NCBI)  COHORT  getz washu  \n",
       "Patient_ID                                                             \n",
       "C3L-00006                                146279.0    UCEC  True   NaN  \n",
       "C3L-00006                                221264.0    UCEC  True  True  \n",
       "C3L-00006                                  2534.0    UCEC  True  True  \n",
       "C3L-00006                                  3910.0    UCEC  True  True  \n",
       "C3L-00006                                  4082.0    UCEC  True  True  \n",
       "\n",
       "[5 rows x 131 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "somatic_mutations = en.get_somatic_mutation('harmonized')\n",
    "somatic_mutations.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Exporting dataframes\n",
    "\n",
    "If you wish to export a dataframe to a file, simply call the dataframe's `to_csv` method, passing the path you wish to save the file to, and the value separator you want:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "clinical = en.get_clinical('mssm')\n",
    "clinical.to_csv(path_or_buf=\"clinical_dataframe.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Downloading data\n",
    "\n",
    "The cptac package stores the data files for each dataset on a remote server. When you first install cptac, you will have no data files. Data files will be automatically downloaded the first time you try to use them. If you won't have internet access and need to download the files for a specific cancer beforehand, that can be done with the \"get\" functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th colspan=\"5\" halign=\"left\">A1BG</th>\n",
       "      <th colspan=\"5\" halign=\"left\">A1BG-AS1</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"8\" halign=\"left\">ZZEF1</th>\n",
       "      <th>hsa-mir-1253</th>\n",
       "      <th>hsa-mir-423</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>ENST00000263100.8</th>\n",
       "      <th>ENST00000595014.1</th>\n",
       "      <th>ENST00000596924.1</th>\n",
       "      <th>ENST00000598345.1</th>\n",
       "      <th>ENST00000600966.1</th>\n",
       "      <th>ENST00000593374.5</th>\n",
       "      <th>ENST00000593960.5</th>\n",
       "      <th>ENST00000594950.5</th>\n",
       "      <th>ENST00000595302.1</th>\n",
       "      <th>ENST00000599728.5</th>\n",
       "      <th>...</th>\n",
       "      <th>ENST00000572426.5</th>\n",
       "      <th>ENST00000572699.1</th>\n",
       "      <th>ENST00000572831.1</th>\n",
       "      <th>ENST00000573183.1</th>\n",
       "      <th>ENST00000573536.1</th>\n",
       "      <th>ENST00000573606.2</th>\n",
       "      <th>ENST00000574474.1</th>\n",
       "      <th>ENST00000575428.1</th>\n",
       "      <th>ENST00000609567.1</th>\n",
       "      <th>ENST00000586878.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Database_ID</th>\n",
       "      <th>ENSG00000121410.12</th>\n",
       "      <th>ENSG00000121410.12</th>\n",
       "      <th>ENSG00000121410.12</th>\n",
       "      <th>ENSG00000121410.12</th>\n",
       "      <th>ENSG00000121410.12</th>\n",
       "      <th>ENSG00000268895.6</th>\n",
       "      <th>ENSG00000268895.6</th>\n",
       "      <th>ENSG00000268895.6</th>\n",
       "      <th>ENSG00000268895.6</th>\n",
       "      <th>ENSG00000268895.6</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000074755.15</th>\n",
       "      <th>ENSG00000074755.15</th>\n",
       "      <th>ENSG00000074755.15</th>\n",
       "      <th>ENSG00000074755.15</th>\n",
       "      <th>ENSG00000074755.15</th>\n",
       "      <th>ENSG00000074755.15</th>\n",
       "      <th>ENSG00000074755.15</th>\n",
       "      <th>ENSG00000074755.15</th>\n",
       "      <th>ENSG00000272920.1</th>\n",
       "      <th>ENSG00000266919.3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>01BR001</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.15</td>\n",
       "      <td>10.13</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.21</td>\n",
       "      <td>1.26</td>\n",
       "      <td>1.28</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.62</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.38</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01BR008</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.06</td>\n",
       "      <td>3.12</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.98</td>\n",
       "      <td>1.94</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.74</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01BR009</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.47</td>\n",
       "      <td>11.88</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.22</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.32</td>\n",
       "      <td>1.19</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.53</td>\n",
       "      <td>1.55</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01BR010</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.25</td>\n",
       "      <td>12.34</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.41</td>\n",
       "      <td>2.05</td>\n",
       "      <td>0.79</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.30</td>\n",
       "      <td>1.66</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.60</td>\n",
       "      <td>1.21</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01BR015</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.39</td>\n",
       "      <td>19.99</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.93</td>\n",
       "      <td>0.79</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.77</td>\n",
       "      <td>1.29</td>\n",
       "      <td>0.48</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2.08</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21BR010</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.20</td>\n",
       "      <td>1.47</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.53</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.38</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.56</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.87</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22BR005</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.21</td>\n",
       "      <td>2.33</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.56</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.54</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.43</td>\n",
       "      <td>1.44</td>\n",
       "      <td>0.72</td>\n",
       "      <td>0.32</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22BR006</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.94</td>\n",
       "      <td>0.21</td>\n",
       "      <td>4.98</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.08</td>\n",
       "      <td>1.34</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.31</td>\n",
       "      <td>1.08</td>\n",
       "      <td>0.85</td>\n",
       "      <td>0.52</td>\n",
       "      <td>3.46</td>\n",
       "      <td>0.67</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPT000814</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>21.64</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.64</td>\n",
       "      <td>2.38</td>\n",
       "      <td>1.05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.49</td>\n",
       "      <td>0.43</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPT001846</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.07</td>\n",
       "      <td>51.13</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.83</td>\n",
       "      <td>2.24</td>\n",
       "      <td>0.65</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121 rows Ã— 225997 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name                        A1BG                                        \\\n",
       "Transcript_ID  ENST00000263100.8  ENST00000595014.1  ENST00000596924.1   \n",
       "Database_ID   ENSG00000121410.12 ENSG00000121410.12 ENSG00000121410.12   \n",
       "Patient_ID                                                               \n",
       "01BR001                      0.0               0.06               0.15   \n",
       "01BR008                      0.0               0.09               0.06   \n",
       "01BR009                      0.0               0.13               0.47   \n",
       "01BR010                      0.0               0.16               0.25   \n",
       "01BR015                      0.0               0.04               0.39   \n",
       "...                          ...                ...                ...   \n",
       "21BR010                      0.0               0.17               0.20   \n",
       "22BR005                      0.0               0.03               0.21   \n",
       "22BR006                      0.0               0.94               0.21   \n",
       "CPT000814                    0.0               0.00               0.00   \n",
       "CPT001846                    0.0               0.05               0.07   \n",
       "\n",
       "Name                                                         A1BG-AS1  \\\n",
       "Transcript_ID  ENST00000598345.1  ENST00000600966.1 ENST00000593374.5   \n",
       "Database_ID   ENSG00000121410.12 ENSG00000121410.12 ENSG00000268895.6   \n",
       "Patient_ID                                                              \n",
       "01BR001                    10.13               0.00               0.0   \n",
       "01BR008                     3.12               0.11               0.0   \n",
       "01BR009                    11.88               0.00               0.0   \n",
       "01BR010                    12.34               0.00               0.0   \n",
       "01BR015                    19.99               0.00               0.0   \n",
       "...                          ...                ...               ...   \n",
       "21BR010                     1.47               0.00               0.0   \n",
       "22BR005                     2.33               0.00               0.0   \n",
       "22BR006                     4.98               0.00               0.0   \n",
       "CPT000814                  21.64               0.17               0.0   \n",
       "CPT001846                  51.13               0.05               0.0   \n",
       "\n",
       "Name                                                                 \\\n",
       "Transcript_ID ENST00000593960.5 ENST00000594950.5 ENST00000595302.1   \n",
       "Database_ID   ENSG00000268895.6 ENSG00000268895.6 ENSG00000268895.6   \n",
       "Patient_ID                                                            \n",
       "01BR001                    0.21              1.26              1.28   \n",
       "01BR008                    0.00              2.98              1.94   \n",
       "01BR009                    0.00              1.22              0.10   \n",
       "01BR010                    0.41              2.05              0.79   \n",
       "01BR015                    0.00              1.93              0.79   \n",
       "...                         ...               ...               ...   \n",
       "21BR010                    0.00              0.00              0.53   \n",
       "22BR005                    0.00              0.00              0.56   \n",
       "22BR006                    0.24              0.08              1.34   \n",
       "CPT000814                  0.64              2.38              1.05   \n",
       "CPT001846                  0.83              2.24              0.65   \n",
       "\n",
       "Name                             ...              ZZEF1                     \\\n",
       "Transcript_ID ENST00000599728.5  ...  ENST00000572426.5  ENST00000572699.1   \n",
       "Database_ID   ENSG00000268895.6  ... ENSG00000074755.15 ENSG00000074755.15   \n",
       "Patient_ID                       ...                                         \n",
       "01BR001                     0.0  ...               0.27               0.62   \n",
       "01BR008                     0.0  ...               0.05               0.12   \n",
       "01BR009                     0.0  ...               0.32               1.19   \n",
       "01BR010                     0.0  ...               0.30               1.66   \n",
       "01BR015                     0.0  ...               0.77               1.29   \n",
       "...                         ...  ...                ...                ...   \n",
       "21BR010                     0.0  ...               1.38               0.56   \n",
       "22BR005                     0.0  ...               0.19               0.54   \n",
       "22BR006                     0.0  ...               1.31               1.08   \n",
       "CPT000814                   0.0  ...               0.00               0.00   \n",
       "CPT001846                   0.0  ...               0.06               0.00   \n",
       "\n",
       "Name                                                                    \\\n",
       "Transcript_ID  ENST00000572831.1  ENST00000573183.1  ENST00000573536.1   \n",
       "Database_ID   ENSG00000074755.15 ENSG00000074755.15 ENSG00000074755.15   \n",
       "Patient_ID                                                               \n",
       "01BR001                     0.00               0.00               1.38   \n",
       "01BR008                     0.00               0.00               0.74   \n",
       "01BR009                     0.00               0.00               3.53   \n",
       "01BR010                     0.00               0.00               1.60   \n",
       "01BR015                     0.48               0.00               2.08   \n",
       "...                          ...                ...                ...   \n",
       "21BR010                     0.00               0.00               1.56   \n",
       "22BR005                     0.00               0.43               1.44   \n",
       "22BR006                     0.85               0.52               3.46   \n",
       "CPT000814                   0.00               0.00               0.49   \n",
       "CPT001846                   0.00               0.13               0.14   \n",
       "\n",
       "Name                                                                    \\\n",
       "Transcript_ID  ENST00000573606.2  ENST00000574474.1  ENST00000575428.1   \n",
       "Database_ID   ENSG00000074755.15 ENSG00000074755.15 ENSG00000074755.15   \n",
       "Patient_ID                                                               \n",
       "01BR001                     0.19               0.14               0.00   \n",
       "01BR008                     0.27               0.03               0.00   \n",
       "01BR009                     1.55               0.32               0.28   \n",
       "01BR010                     1.21               0.10               0.00   \n",
       "01BR015                     0.50               0.36               0.00   \n",
       "...                          ...                ...                ...   \n",
       "21BR010                     0.95               0.87               0.22   \n",
       "22BR005                     0.72               0.32               0.00   \n",
       "22BR006                     0.67               0.29               0.00   \n",
       "CPT000814                   0.43               0.05               0.00   \n",
       "CPT001846                   0.00               0.02               0.00   \n",
       "\n",
       "Name               hsa-mir-1253       hsa-mir-423  \n",
       "Transcript_ID ENST00000609567.1 ENST00000586878.1  \n",
       "Database_ID   ENSG00000272920.1 ENSG00000266919.3  \n",
       "Patient_ID                                         \n",
       "01BR001                     0.0               0.0  \n",
       "01BR008                     0.0               0.0  \n",
       "01BR009                     0.0               0.0  \n",
       "01BR010                     0.0               0.0  \n",
       "01BR015                     0.0               0.0  \n",
       "...                         ...               ...  \n",
       "21BR010                     0.0               0.0  \n",
       "22BR005                     0.0               0.0  \n",
       "22BR006                     0.0               0.0  \n",
       "CPT000814                   0.0               0.0  \n",
       "CPT001846                   0.0               0.0  \n",
       "\n",
       "[121 rows x 225997 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "br = cptac.Brca()\n",
    "br.get_transcriptomics('broad')\n",
    "#This downloads and displays the brca transcriptomic data from the \"broad\" source"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting help with a dataset or function\n",
    "\n",
    "To view the documentation for a dataset, pass it to the Python `help` function, e.g. `help(en)`. You can also view the documentation for just a specific function: `help(en.join_omics_to_omics)`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method join_omics_to_omics in module cptac.cancers.cancer:\n",
      "\n",
      "join_omics_to_omics(df1_name: str, df2_name: str, df1_source: str = None, df2_source: str = None, genes1: str = None, genes2: str = None, how: str = 'outer', quiet: bool = False, tissue_type: str = 'both') -> pandas.core.frame.DataFrame method of cptac.cancers.ucec.Ucec instance\n",
      "    Take specified column(s) from one omics dataframe, and join to specified columns(s) from another omics dataframe. Intersection (inner join) of indices is used.\n",
      "    \n",
      "    Parameters:\n",
      "    df1_name (str): Name of first omics dataframe to select columns from.\n",
      "    df2_name (str): Name of second omics dataframe to select columns from.\n",
      "    df1_source (str): Name of source for the first omics dataframe.\n",
      "    df2_source (str): Name of source for the second omics dataframe.\n",
      "    genes1 (str, or list or array-like of str, optional): Gene(s) for column(s) to select from df1_name. str if one key, list or array-like of str if multiple.\n",
      "        Default of None will select entire dataframe.\n",
      "    genes2 (str, or list or array-like of str, optional): Gene(s) for Column(s) to select from df2_name. str if one key, list or array-like of str if multiple.\n",
      "        Default of None will select entire dataframe.\n",
      "    how (str, optional): How to perform the join, acceptable values are from ['outer', 'inner', 'left', 'right']. Defaults to 'outer'.\n",
      "    quiet (bool, optional): Whether to warn when inserting NaNs. Defaults to False.\n",
      "    tissue_type (str): Acceptable values in [\"tumor\",\"normal\",\"both\"]. Specifies the desired tissue type for the dataframe. Defaults to \"both\".\n",
      "    \n",
      "    Returns:\n",
      "    pandas.DataFrame: The selected columns from the two omics dataframes, joined into one dataframe.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(en.join_omics_to_omics)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
