{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial 4: Multi-level column indices (`MultiIndex`)\n",
    "\n",
    "A MultiIndex is an index with multiple, hierarchical levels. We use a multiindex for the column headers of a \n",
    "dataframe. This allows each column to have multiple keys associated with it--one key for each level. \n",
    "\n",
    "The multiindex is a native part of the `pandas` package. For more documentation, see their [MultiIndex / advanced indexing page](https://pandas.pydata.org/pandas-docs/stable/user_guide/advanced.html).\n",
    "\n",
    "MultiIndexes make it easy to associate multiple identifiers with a single column.  To show what this looks like we'll look at phosphoproteomics data in the endometrial cancer dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Data type</th>\n",
       "      <th>Available sources</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CNV</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>circular_RNA</td>\n",
       "      <td>bcm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>clinical</td>\n",
       "      <td>mssm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>deconvolution_cibersort</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>deconvolution_xcell</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>follow-up</td>\n",
       "      <td>mssm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>medical_history</td>\n",
       "      <td>mssm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>miRNA</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>phosphoproteomics</td>\n",
       "      <td>umich</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>proteomics</td>\n",
       "      <td>umich</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>somatic_mutation</td>\n",
       "      <td>harmonized, washu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>bcm, broad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>tumor_purity</td>\n",
       "      <td>washu</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Data type  Available sources\n",
       "0                       CNV              washu\n",
       "1              circular_RNA                bcm\n",
       "2                  clinical               mssm\n",
       "3   deconvolution_cibersort              washu\n",
       "4       deconvolution_xcell              washu\n",
       "5                 follow-up               mssm\n",
       "6           medical_history               mssm\n",
       "7                     miRNA              washu\n",
       "8         phosphoproteomics              umich\n",
       "9                proteomics              umich\n",
       "10         somatic_mutation  harmonized, washu\n",
       "11          transcriptomics         bcm, broad\n",
       "12             tumor_purity              washu"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Throughout the tutorial we will be using the endometrial and colon datasets\n",
    "import cptac\n",
    "#cptac.download_cancer('ucec')\n",
    "en = cptac.Ucec()\n",
    "en.list_data_sources()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th>ARF5</th>\n",
       "      <th>M6PR</th>\n",
       "      <th colspan=\"8\" halign=\"left\">ESRRA</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SCRIB</th>\n",
       "      <th colspan=\"6\" halign=\"left\">TSGA10</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SVIL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Site</th>\n",
       "      <th>S137</th>\n",
       "      <th>S267</th>\n",
       "      <th>S19</th>\n",
       "      <th>S22</th>\n",
       "      <th>S19S22</th>\n",
       "      <th>T31</th>\n",
       "      <th>S19S22</th>\n",
       "      <th>S19S22S27</th>\n",
       "      <th>S19S22T31</th>\n",
       "      <th>S27</th>\n",
       "      <th>...</th>\n",
       "      <th>S1575T1588S1594</th>\n",
       "      <th>S1594</th>\n",
       "      <th>S11</th>\n",
       "      <th>S173</th>\n",
       "      <th>S213</th>\n",
       "      <th>S391</th>\n",
       "      <th>S779</th>\n",
       "      <th>S101</th>\n",
       "      <th>S296</th>\n",
       "      <th>S459</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peptide</th>\n",
       "      <th>QDMPNAMPVsELTDK</th>\n",
       "      <th>GVGDDQLGEEsEERDDHLLPM</th>\n",
       "      <th>AEPAsPDSPK</th>\n",
       "      <th>AEPASPDsPK</th>\n",
       "      <th>AEPAsPDsPK</th>\n",
       "      <th>AEPASPDSPKGSSETEtEPPVALAPGPAPTR</th>\n",
       "      <th>AEPAsPDsPKGSSETETEPPVALAPGPAPTR</th>\n",
       "      <th>AEPAsPDsPKGSsETETEPPVALAPGPAPTR</th>\n",
       "      <th>AEPAsPDsPKGSSETEtEPPVALAPGPAPTR</th>\n",
       "      <th>GSsETETEPPVALAPGPAPTR</th>\n",
       "      <th>...</th>\n",
       "      <th>LAEAPSPAPTPsPTPVEDLGPQTStSPGRLsPDFAEELR</th>\n",
       "      <th>LsPDFAEELR</th>\n",
       "      <th>sPGRDPELQVEAAEVTTK</th>\n",
       "      <th>sPSRLDSFVK</th>\n",
       "      <th>RPsPTAR</th>\n",
       "      <th>AMDTEsELGR</th>\n",
       "      <th>GLDRsLEENLCYR;GLDRsLEENLCYRDF</th>\n",
       "      <th>EVVSSQVDDLTsHNEHLCK</th>\n",
       "      <th>DSEGDTPsLINWPSSK</th>\n",
       "      <th>LPsPTVAR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Database_ID</th>\n",
       "      <th>ENSP00000000233.5</th>\n",
       "      <th>ENSP00000000412.3</th>\n",
       "      <th>ENSP00000000442.6</th>\n",
       "      <th>ENSP00000000442.6</th>\n",
       "      <th>ENSP00000000442.6</th>\n",
       "      <th>ENSP00000000442.6</th>\n",
       "      <th>ENSP00000000442.6</th>\n",
       "      <th>ENSP00000000442.6</th>\n",
       "      <th>ENSP00000000442.6</th>\n",
       "      <th>ENSP00000000442.6</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSP00000501177.1</th>\n",
       "      <th>ENSP00000501177.1</th>\n",
       "      <th>ENSP00000501312.1</th>\n",
       "      <th>ENSP00000501312.1</th>\n",
       "      <th>ENSP00000501312.1</th>\n",
       "      <th>ENSP00000501312.1</th>\n",
       "      <th>ENSP00000501312.1</th>\n",
       "      <th>ENSP00000501312.1</th>\n",
       "      <th>ENSP00000501521.1</th>\n",
       "      <th>ENSP00000501521.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.573633</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.304721</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.667426</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.905606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.069911</td>\n",
       "      <td>-0.584774</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.561657</td>\n",
       "      <td>-0.652457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00008</th>\n",
       "      <td>0.003632</td>\n",
       "      <td>-0.393734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.789193</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.488427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.431599</td>\n",
       "      <td>-1.079638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00032</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.211020</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.131605</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.104862</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.439041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00084</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.220473</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.290506</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.399718</td>\n",
       "      <td>-0.875016</td>\n",
       "      <td>-0.579824</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.505807</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.521725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00090</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.161496</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.708453</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.405402</td>\n",
       "      <td>1.253045</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.265813</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.069439</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.510268</td>\n",
       "      <td>-1.889144</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.592203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.126482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NX5.N</th>\n",
       "      <td>-0.311416</td>\n",
       "      <td>0.211987</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.395850</td>\n",
       "      <td>0.045874</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.359392</td>\n",
       "      <td>-0.046312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.019187</td>\n",
       "      <td>-2.539302</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.965335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NX6.N</th>\n",
       "      <td>0.001346</td>\n",
       "      <td>0.513387</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.335033</td>\n",
       "      <td>-0.108266</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.439623</td>\n",
       "      <td>-0.211645</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.862658</td>\n",
       "      <td>-0.727766</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.339839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NX7.N</th>\n",
       "      <td>0.186926</td>\n",
       "      <td>0.141754</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.106618</td>\n",
       "      <td>-0.313644</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.577533</td>\n",
       "      <td>-0.622015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.932881</td>\n",
       "      <td>-2.165876</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.115747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NX8.N</th>\n",
       "      <td>0.227515</td>\n",
       "      <td>0.120051</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.059253</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.149488</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.241592</td>\n",
       "      <td>-0.993248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NX9.N</th>\n",
       "      <td>0.183929</td>\n",
       "      <td>0.171886</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.088318</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.799354</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.990747</td>\n",
       "      <td>0.422213</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>152 rows × 85661 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name                     ARF5                  M6PR             ESRRA  \\\n",
       "Site                     S137                  S267               S19   \n",
       "Peptide       QDMPNAMPVsELTDK GVGDDQLGEEsEERDDHLLPM        AEPAsPDSPK   \n",
       "Database_ID ENSP00000000233.5     ENSP00000000412.3 ENSP00000000442.6   \n",
       "Patient_ID                                                              \n",
       "C3L-00006                 NaN              0.573633               NaN   \n",
       "C3L-00008            0.003632             -0.393734               NaN   \n",
       "C3L-00032                 NaN             -0.211020               NaN   \n",
       "C3L-00084                 NaN              0.220473               NaN   \n",
       "C3L-00090                 NaN              0.161496               NaN   \n",
       "...                       ...                   ...               ...   \n",
       "NX5.N               -0.311416              0.211987               NaN   \n",
       "NX6.N                0.001346              0.513387               NaN   \n",
       "NX7.N                0.186926              0.141754               NaN   \n",
       "NX8.N                0.227515              0.120051               NaN   \n",
       "NX9.N                0.183929              0.171886               NaN   \n",
       "\n",
       "Name                                             \\\n",
       "Site                      S22            S19S22   \n",
       "Peptide            AEPASPDsPK        AEPAsPDsPK   \n",
       "Database_ID ENSP00000000442.6 ENSP00000000442.6   \n",
       "Patient_ID                                        \n",
       "C3L-00006                 NaN          0.304721   \n",
       "C3L-00008                 NaN          0.789193   \n",
       "C3L-00032                 NaN               NaN   \n",
       "C3L-00084                 NaN         -0.290506   \n",
       "C3L-00090                 NaN          0.708453   \n",
       "...                       ...               ...   \n",
       "NX5.N                0.395850          0.045874   \n",
       "NX6.N               -0.335033         -0.108266   \n",
       "NX7.N               -0.106618         -0.313644   \n",
       "NX8.N                     NaN         -0.059253   \n",
       "NX9.N                     NaN          0.088318   \n",
       "\n",
       "Name                                                                         \\\n",
       "Site                                    T31                          S19S22   \n",
       "Peptide     AEPASPDSPKGSSETEtEPPVALAPGPAPTR AEPAsPDsPKGSSETETEPPVALAPGPAPTR   \n",
       "Database_ID               ENSP00000000442.6               ENSP00000000442.6   \n",
       "Patient_ID                                                                    \n",
       "C3L-00006                               NaN                             NaN   \n",
       "C3L-00008                               NaN                             NaN   \n",
       "C3L-00032                               NaN                             NaN   \n",
       "C3L-00084                               NaN                             NaN   \n",
       "C3L-00090                               NaN                        0.405402   \n",
       "...                                     ...                             ...   \n",
       "NX5.N                                   NaN                             NaN   \n",
       "NX6.N                                   NaN                             NaN   \n",
       "NX7.N                                   NaN                             NaN   \n",
       "NX8.N                                   NaN                             NaN   \n",
       "NX9.N                                   NaN                             NaN   \n",
       "\n",
       "Name                                                                         \\\n",
       "Site                              S19S22S27                       S19S22T31   \n",
       "Peptide     AEPAsPDsPKGSsETETEPPVALAPGPAPTR AEPAsPDsPKGSSETEtEPPVALAPGPAPTR   \n",
       "Database_ID               ENSP00000000442.6               ENSP00000000442.6   \n",
       "Patient_ID                                                                    \n",
       "C3L-00006                               NaN                             NaN   \n",
       "C3L-00008                               NaN                             NaN   \n",
       "C3L-00032                               NaN                             NaN   \n",
       "C3L-00084                               NaN                             NaN   \n",
       "C3L-00090                          1.253045                             NaN   \n",
       "...                                     ...                             ...   \n",
       "NX5.N                                   NaN                             NaN   \n",
       "NX6.N                                   NaN                             NaN   \n",
       "NX7.N                                   NaN                             NaN   \n",
       "NX8.N                                   NaN                             NaN   \n",
       "NX9.N                                   NaN                             NaN   \n",
       "\n",
       "Name                               ...  \\\n",
       "Site                          S27  ...   \n",
       "Peptide     GSsETETEPPVALAPGPAPTR  ...   \n",
       "Database_ID     ENSP00000000442.6  ...   \n",
       "Patient_ID                         ...   \n",
       "C3L-00006                     NaN  ...   \n",
       "C3L-00008                     NaN  ...   \n",
       "C3L-00032                0.131605  ...   \n",
       "C3L-00084                     NaN  ...   \n",
       "C3L-00090                0.265813  ...   \n",
       "...                           ...  ...   \n",
       "NX5.N                         NaN  ...   \n",
       "NX6.N                         NaN  ...   \n",
       "NX7.N                         NaN  ...   \n",
       "NX8.N                         NaN  ...   \n",
       "NX9.N                         NaN  ...   \n",
       "\n",
       "Name                                          SCRIB                    \\\n",
       "Site                                S1575T1588S1594             S1594   \n",
       "Peptide     LAEAPSPAPTPsPTPVEDLGPQTStSPGRLsPDFAEELR        LsPDFAEELR   \n",
       "Database_ID                       ENSP00000501177.1 ENSP00000501177.1   \n",
       "Patient_ID                                                              \n",
       "C3L-00006                                       NaN          0.667426   \n",
       "C3L-00008                                       NaN               NaN   \n",
       "C3L-00032                                       NaN          0.104862   \n",
       "C3L-00084                                       NaN          0.399718   \n",
       "C3L-00090                                       NaN          1.069439   \n",
       "...                                             ...               ...   \n",
       "NX5.N                                      1.359392         -0.046312   \n",
       "NX6.N                                     -0.439623         -0.211645   \n",
       "NX7.N                                      1.577533         -0.622015   \n",
       "NX8.N                                           NaN               NaN   \n",
       "NX9.N                                           NaN               NaN   \n",
       "\n",
       "Name                    TSGA10                                      \\\n",
       "Site                       S11              S173              S213   \n",
       "Peptide     sPGRDPELQVEAAEVTTK        sPSRLDSFVK           RPsPTAR   \n",
       "Database_ID  ENSP00000501312.1 ENSP00000501312.1 ENSP00000501312.1   \n",
       "Patient_ID                                                           \n",
       "C3L-00006                  NaN          0.905606               NaN   \n",
       "C3L-00008                  NaN         -0.488427               NaN   \n",
       "C3L-00032                  NaN               NaN               NaN   \n",
       "C3L-00084            -0.875016         -0.579824               NaN   \n",
       "C3L-00090                  NaN          0.510268         -1.889144   \n",
       "...                        ...               ...               ...   \n",
       "NX5.N                      NaN         -1.019187         -2.539302   \n",
       "NX6.N                      NaN         -0.862658         -0.727766   \n",
       "NX7.N                      NaN         -0.932881         -2.165876   \n",
       "NX8.N                      NaN         -0.149488               NaN   \n",
       "NX9.N                      NaN         -0.799354               NaN   \n",
       "\n",
       "Name                                                         \\\n",
       "Site                     S391                          S779   \n",
       "Peptide            AMDTEsELGR GLDRsLEENLCYR;GLDRsLEENLCYRDF   \n",
       "Database_ID ENSP00000501312.1             ENSP00000501312.1   \n",
       "Patient_ID                                                    \n",
       "C3L-00006           -0.069911                     -0.584774   \n",
       "C3L-00008                 NaN                           NaN   \n",
       "C3L-00032                 NaN                           NaN   \n",
       "C3L-00084                 NaN                     -0.505807   \n",
       "C3L-00090                 NaN                     -0.592203   \n",
       "...                       ...                           ...   \n",
       "NX5.N                     NaN                           NaN   \n",
       "NX6.N                     NaN                           NaN   \n",
       "NX7.N                     NaN                           NaN   \n",
       "NX8.N                     NaN                           NaN   \n",
       "NX9.N                     NaN                           NaN   \n",
       "\n",
       "Name                                         SVIL                    \n",
       "Site                       S101              S296              S459  \n",
       "Peptide     EVVSSQVDDLTsHNEHLCK  DSEGDTPsLINWPSSK          LPsPTVAR  \n",
       "Database_ID   ENSP00000501312.1 ENSP00000501521.1 ENSP00000501521.1  \n",
       "Patient_ID                                                           \n",
       "C3L-00006                   NaN         -0.561657         -0.652457  \n",
       "C3L-00008                   NaN         -0.431599         -1.079638  \n",
       "C3L-00032                   NaN               NaN         -1.439041  \n",
       "C3L-00084                   NaN               NaN         -1.521725  \n",
       "C3L-00090                   NaN               NaN         -1.126482  \n",
       "...                         ...               ...               ...  \n",
       "NX5.N                       NaN               NaN         -0.965335  \n",
       "NX6.N                       NaN               NaN         -0.339839  \n",
       "NX7.N                       NaN               NaN          0.115747  \n",
       "NX8.N                       NaN         -0.241592         -0.993248  \n",
       "NX9.N                       NaN          0.990747          0.422213  \n",
       "\n",
       "[152 rows x 85661 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "phospho = en.get_proteomics('umich')\n",
    "en.get_phosphoproteomics('umich')\n",
    "#en.get_CNV('washu')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first level of keys is the name of the gene associated with the column of data. The second level is the site of acetylation. Each column has its own Name and Site associated with it. \n",
    "Normally, as above, keys that repeat for adjacent columns are not displayed, to make it easier to read the dataframe. For example, in the table above the first 7 columns are all for the A2M gene, but the label is only shown for the first column. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Join functions with multiindices \n",
    "The join functions have been written to handle multiindices. More information on the join functions can be found in the joining_dataframes tutorial. \n",
    "An example of joining a multiindexed dataframe (in this case phosphoproteomics) with a non multiindexed dataframe (in this case CNV) is below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DATATYPE: CNV\n",
      "source: washu\n",
      "LOAD FUNCTIONS: {'somatic_mutation': <bound method WashuUcec.load_somatic_mutation of <cptac.cancers.washu.washuucec.WashuUcec object at 0x7ff6d4214e80>>, 'miRNA': <bound method WashuUcec.load_miRNA of <cptac.cancers.washu.washuucec.WashuUcec object at 0x7ff6d4214e80>>, 'xcell': <bound method WashuUcec.load_xcell of <cptac.cancers.washu.washuucec.WashuUcec object at 0x7ff6d4214e80>>, 'cibersort': <bound method WashuUcec.load_cibersort of <cptac.cancers.washu.washuucec.WashuUcec object at 0x7ff6d4214e80>>, 'CNV': <bound method WashuUcec.load_CNV of <cptac.cancers.washu.washuucec.WashuUcec object at 0x7ff6d4214e80>>, 'tumor_purity': <bound method WashuUcec.load_tumor_purity of <cptac.cancers.washu.washuucec.WashuUcec object at 0x7ff6d4214e80>>}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Polars found a filename. Ensure you pass a path to the file instead of a python file object when possible for best performance."
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DATATYPE: phosphoproteomics\n",
      "source: umich\n",
      "LOAD FUNCTIONS: {'phosphoproteomics': <bound method UmichUcec.load_phosphoproteomics of <cptac.cancers.umich.umichucec.UmichUcec object at 0x7ff6d4214ca0>>, 'proteomics': <bound method UmichUcec.load_proteomics of <cptac.cancers.umich.umichucec.UmichUcec object at 0x7ff6d4214ca0>>}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th>A1BG_washu_CNV</th>\n",
       "      <th>A1CF_washu_CNV</th>\n",
       "      <th>A2M_washu_CNV</th>\n",
       "      <th>A2ML1_washu_CNV</th>\n",
       "      <th>A3GALT2_washu_CNV</th>\n",
       "      <th>A4GALT_washu_CNV</th>\n",
       "      <th>A4GNT_washu_CNV</th>\n",
       "      <th>AAAS_washu_CNV</th>\n",
       "      <th>AACS_washu_CNV</th>\n",
       "      <th>AADAC_washu_CNV</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SCRIB_umich_phosphoproteomics</th>\n",
       "      <th colspan=\"6\" halign=\"left\">TSGA10_umich_phosphoproteomics</th>\n",
       "      <th colspan=\"2\" halign=\"left\">SVIL_umich_phosphoproteomics</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Site</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>...</th>\n",
       "      <th>S1575T1588S1594</th>\n",
       "      <th>S1594</th>\n",
       "      <th>S11</th>\n",
       "      <th>S173</th>\n",
       "      <th>S213</th>\n",
       "      <th>S391</th>\n",
       "      <th>S779</th>\n",
       "      <th>S101</th>\n",
       "      <th>S296</th>\n",
       "      <th>S459</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peptide</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>...</th>\n",
       "      <th>LAEAPSPAPTPsPTPVEDLGPQTStSPGRLsPDFAEELR</th>\n",
       "      <th>LsPDFAEELR</th>\n",
       "      <th>sPGRDPELQVEAAEVTTK</th>\n",
       "      <th>sPSRLDSFVK</th>\n",
       "      <th>RPsPTAR</th>\n",
       "      <th>AMDTEsELGR</th>\n",
       "      <th>GLDRsLEENLCYR;GLDRsLEENLCYRDF</th>\n",
       "      <th>EVVSSQVDDLTsHNEHLCK</th>\n",
       "      <th>DSEGDTPsLINWPSSK</th>\n",
       "      <th>LPsPTVAR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3L-00006</th>\n",
       "      <td>-0.00659</td>\n",
       "      <td>-0.01982</td>\n",
       "      <td>-0.01402</td>\n",
       "      <td>-0.01402</td>\n",
       "      <td>-0.01418</td>\n",
       "      <td>-0.00839</td>\n",
       "      <td>-0.01305</td>\n",
       "      <td>-0.01402</td>\n",
       "      <td>-0.01402</td>\n",
       "      <td>-0.01305</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.667426</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.905606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.069911</td>\n",
       "      <td>-0.584774</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.561657</td>\n",
       "      <td>-0.652457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00008</th>\n",
       "      <td>0.02578</td>\n",
       "      <td>0.00726</td>\n",
       "      <td>0.01350</td>\n",
       "      <td>0.01350</td>\n",
       "      <td>0.00732</td>\n",
       "      <td>0.01642</td>\n",
       "      <td>0.01005</td>\n",
       "      <td>0.01225</td>\n",
       "      <td>0.01225</td>\n",
       "      <td>0.01005</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.488427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.431599</td>\n",
       "      <td>-1.079638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00032</th>\n",
       "      <td>0.01262</td>\n",
       "      <td>0.00425</td>\n",
       "      <td>-0.00275</td>\n",
       "      <td>-0.00275</td>\n",
       "      <td>0.00166</td>\n",
       "      <td>0.00549</td>\n",
       "      <td>-0.00038</td>\n",
       "      <td>-0.00275</td>\n",
       "      <td>-0.00275</td>\n",
       "      <td>-0.00038</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.104862</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.439041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00084</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.399718</td>\n",
       "      <td>-0.875016</td>\n",
       "      <td>-0.579824</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.505807</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.521725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3L-00090</th>\n",
       "      <td>0.00100</td>\n",
       "      <td>0.41191</td>\n",
       "      <td>-0.02299</td>\n",
       "      <td>-0.02299</td>\n",
       "      <td>-0.02436</td>\n",
       "      <td>-0.01198</td>\n",
       "      <td>-0.03307</td>\n",
       "      <td>-0.02299</td>\n",
       "      <td>-0.02299</td>\n",
       "      <td>-0.03307</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.069439</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.510268</td>\n",
       "      <td>-1.889144</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.592203</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.126482</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 104580 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       A1BG_washu_CNV A1CF_washu_CNV A2M_washu_CNV A2ML1_washu_CNV  \\\n",
       "Site                                                                     \n",
       "Peptide                                                                  \n",
       "Patient_ID                                                               \n",
       "C3L-00006        -0.00659       -0.01982      -0.01402        -0.01402   \n",
       "C3L-00008         0.02578        0.00726       0.01350         0.01350   \n",
       "C3L-00032         0.01262        0.00425      -0.00275        -0.00275   \n",
       "C3L-00084             NaN            NaN           NaN             NaN   \n",
       "C3L-00090         0.00100        0.41191      -0.02299        -0.02299   \n",
       "\n",
       "Name       A3GALT2_washu_CNV A4GALT_washu_CNV A4GNT_washu_CNV AAAS_washu_CNV  \\\n",
       "Site                                                                           \n",
       "Peptide                                                                        \n",
       "Patient_ID                                                                     \n",
       "C3L-00006           -0.01418         -0.00839        -0.01305       -0.01402   \n",
       "C3L-00008            0.00732          0.01642         0.01005        0.01225   \n",
       "C3L-00032            0.00166          0.00549        -0.00038       -0.00275   \n",
       "C3L-00084                NaN              NaN             NaN            NaN   \n",
       "C3L-00090           -0.02436         -0.01198        -0.03307       -0.02299   \n",
       "\n",
       "Name       AACS_washu_CNV AADAC_washu_CNV  ...  \\\n",
       "Site                                       ...   \n",
       "Peptide                                    ...   \n",
       "Patient_ID                                 ...   \n",
       "C3L-00006        -0.01402        -0.01305  ...   \n",
       "C3L-00008         0.01225         0.01005  ...   \n",
       "C3L-00032        -0.00275        -0.00038  ...   \n",
       "C3L-00084             NaN             NaN  ...   \n",
       "C3L-00090        -0.02299        -0.03307  ...   \n",
       "\n",
       "Name                 SCRIB_umich_phosphoproteomics             \\\n",
       "Site                               S1575T1588S1594      S1594   \n",
       "Peptide    LAEAPSPAPTPsPTPVEDLGPQTStSPGRLsPDFAEELR LsPDFAEELR   \n",
       "Patient_ID                                                      \n",
       "C3L-00006                                      NaN   0.667426   \n",
       "C3L-00008                                      NaN        NaN   \n",
       "C3L-00032                                      NaN   0.104862   \n",
       "C3L-00084                                      NaN   0.399718   \n",
       "C3L-00090                                      NaN   1.069439   \n",
       "\n",
       "Name       TSGA10_umich_phosphoproteomics                                  \\\n",
       "Site                                  S11       S173      S213       S391   \n",
       "Peptide                sPGRDPELQVEAAEVTTK sPSRLDSFVK   RPsPTAR AMDTEsELGR   \n",
       "Patient_ID                                                                  \n",
       "C3L-00006                             NaN   0.905606       NaN  -0.069911   \n",
       "C3L-00008                             NaN  -0.488427       NaN        NaN   \n",
       "C3L-00032                             NaN        NaN       NaN        NaN   \n",
       "C3L-00084                       -0.875016  -0.579824       NaN        NaN   \n",
       "C3L-00090                             NaN   0.510268 -1.889144        NaN   \n",
       "\n",
       "Name                                                          \\\n",
       "Site                                S779                S101   \n",
       "Peptide    GLDRsLEENLCYR;GLDRsLEENLCYRDF EVVSSQVDDLTsHNEHLCK   \n",
       "Patient_ID                                                     \n",
       "C3L-00006                      -0.584774                 NaN   \n",
       "C3L-00008                            NaN                 NaN   \n",
       "C3L-00032                            NaN                 NaN   \n",
       "C3L-00084                      -0.505807                 NaN   \n",
       "C3L-00090                      -0.592203                 NaN   \n",
       "\n",
       "Name       SVIL_umich_phosphoproteomics            \n",
       "Site                               S296      S459  \n",
       "Peptide                DSEGDTPsLINWPSSK  LPsPTVAR  \n",
       "Patient_ID                                         \n",
       "C3L-00006                     -0.561657 -0.652457  \n",
       "C3L-00008                     -0.431599 -1.079638  \n",
       "C3L-00032                           NaN -1.439041  \n",
       "C3L-00084                           NaN -1.521725  \n",
       "C3L-00090                           NaN -1.126482  \n",
       "\n",
       "[5 rows x 104580 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phospho_and_CNV = en.join_omics_to_omics(df1_name=\"CNV\", df2_name=\"phosphoproteomics\", df1_source='washu', df2_source = 'umich')\n",
    "phospho_and_CNV.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since the CNV dataframe doesn't have the level \"Site\", it is filled in with NANs, so that it can be joined to the phosphoproteomics dataframe. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# How to select from multiindex\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Selecting based on all levels\n",
    "We can select single columns by passing the proper keys for all levels of the multiindex. For example, to get the acetylproteomics for A2M at site K1176, we'd do the following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Ucec' object has no attribute 'get_acetylproteomics'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/n2/95tlbmw5173fm8yjh4ffxggr0000gn/T/ipykernel_41378/2971020525.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0macetyl\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0men\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_acetylproteomics\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mall_levels_selection\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0macetyl\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"A2M\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"K1176\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mall_levels_selection\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Ucec' object has no attribute 'get_acetylproteomics'"
     ]
    }
   ],
   "source": [
    "acetyl = en.get_acetylproteomics()\n",
    "all_levels_selection = acetyl[\"A2M\"][\"K1176\"]\n",
    "all_levels_selection.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Selecting based on one level\n",
    "We can easily select multiple columns from our multiindex dataframe, based on just the \"Name\" level of the multiindex:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene1_filter = acetyl.columns.get_level_values(\"Name\").str.startswith(\"AA\") # Select all columns where the gene starts with \"AA\". This will grab every column where the key \"Name\" starts with AA\n",
    "gene1_data = acetyl.loc[:, gene1_filter]\n",
    "gene1_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Selecting based on a different level of the multiindex\n",
    "We can also select based on one of the inner levels of the multiindex. For example, to get data for all tyrosine phosphorylation sites:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_site_filter = phospho.columns.get_level_values(\"Site\").str.contains(\"Y\") # Create a boolean filter selecting all columns where the Site level contains a \"Y\"\n",
    "\n",
    "y_sites = phospho.loc[:, y_site_filter] # Select the columns\n",
    "y_sites.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# How to use `cptac.utils.reduce_multiindex()`\n",
    "To make it easier to work with multi-level indices, we provide the `reduce_multiindex` function, available for import from the `cptac.utils` submodule. It can both drop levels from a multiindex, and \"flatten\" a multi-level index into a single-level index by concatenating the keys from multiple levels into a single key for each column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cptac.utils as ut"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dropping Levels\n",
    "We can drop levels based on index or name. We can also drop single or multiple levels at once. \n",
    "Note that it will warn you if duplicate column key combinations arise due to dropping levels. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dropping by index or name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ut.reduce_multiindex(df=phospho, levels_to_drop=\"Site\").head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dropping single or multiple levels at once\n",
    "By passing a list (or array-like) to levels_to drop, we can drop multiple levels of the multiindex at the same time. Note that we must leave at least one existing level. \n",
    "\n",
    "We will show this with the colon data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colon = cptac.Colon()\n",
    "phospho = colon.get_phosphoproteomics()\n",
    "phospho.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop evey level except Database_ID\n",
    "drop = [\"Name\", \"Site\"]\n",
    "ut.reduce_multiindex(df=phospho, levels_to_drop=drop).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combining levels (Flattening)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can combine levels of a multiindexed dataframe. When combined the levels will be sepereated by an underscore, by default. We could specify a different seperator using the `sep` parameter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ut.reduce_multiindex(df=phospho, flatten=True).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When flatteing levels , NaNs and empty strings will automitically be dropped."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phospho_and_CNV = en.join_omics_to_omics(df1_name=\"CNV\", df2_name=\"phosphoproteomics\")\n",
    "phospho_and_CNV.head()\n",
    "\n",
    "# Note that the CNV columns all have empty strings in the \"Site\" level of the columns,\n",
    "# since the CNV data doesn't have any values for that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ut.reduce_multiindex(df=phospho_and_CNV, flatten=True).head()\n",
    "# Notice that the empty strings have been dropped"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting a single level index of tuples\n",
    "\n",
    "You can also use `reduce_multiindex` to turn the multi-level column index into a single level index of tuples, with each value in a column's tuple corresponding to the column's value for that level of the index:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ut.reduce_multiindex(df=phospho, tuples=True).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Turning off warnings\n",
    "\n",
    "If your multiindex operation creates duplicate column headers, or has no effect, `reduce_multiindex` will warn you. You can silence these warnings by passing `True` to the `quiet` parameter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ut.reduce_multiindex(df=phospho, levels_to_drop=[\"Name\", \"Database_ID\"]).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# No warning will be issued\n",
    "ut.reduce_multiindex(df=phospho, levels_to_drop=[\"Name\", \"Database_ID\"], quiet=True).head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
